README.md
# The Official Rust Connector for [TDengine]

| Docs.rs                                        | Crates.io Version                                  | Crates.io Downloads                                | CodeCov                                                                                                                                                           |
| ---------------------------------------------- | -------------------------------------------------- | -------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| ![docs.rs](https://img.shields.io/docsrs/taos) | ![Crates.io](https://img.shields.io/crates/v/taos) | ![Crates.io](https://img.shields.io/crates/d/taos) | [![codecov](https://codecov.io/gh/taosdata/taos-connector-rust/branch/main/graph/badge.svg?token=P11UKNLTVO)](https://codecov.io/gh/taosdata/taos-connector-rust) |

This is the official TDengine connector in Rust.

## Dependencies

- [Rust](https://www.rust-lang.org/learn/get-started) of course.

if you use the default features, it'll depend on:

- [TDengine] Client library and headers.

## Usage

By default, enable both native and websocket client:

```toml
[dependencies]
taos = "*"
```

For websocket client only:

```toml
[dependencies]
taos = { version = "*", default-features = false, features = ["ws"] }
```

For native only:

```toml
[dependencies]
taos = { version = "*", default-features = false, features = ["native"] }
```

### Query

```rust
use chrono::{DateTime, Local};
use taos::*;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let dsn = "taos://localhost:6030";
    let builder = TaosBuilder::from_dsn(dsn)?;

    let taos = builder.build()?;

    let db = "query";

    // prepare database
    taos.exec_many([
        format!("DROP DATABASE IF EXISTS `{db}`"),
        format!("CREATE DATABASE `{db}`"),
        format!("USE `{db}`"),
    ])
    .await?;

    let inserted = taos.exec_many([
        // create super table
        "CREATE TABLE `meters` (`ts` TIMESTAMP, `current` FLOAT, `voltage` INT, `phase` FLOAT) \
         TAGS (`groupid` INT, `location` BINARY(16))",
        // create child table
        "CREATE TABLE `d0` USING `meters` TAGS(0, 'Los Angles')",
        // insert into child table
        "INSERT INTO `d0` values(now - 10s, 10, 116, 0.32)",
        // insert with NULL values
        "INSERT INTO `d0` values(now - 8s, NULL, NULL, NULL)",
        // insert and automatically create table with tags if not exists
        "INSERT INTO `d1` USING `meters` TAGS(1, 'San Francisco') values(now - 9s, 10.1, 119, 0.33)",
        // insert many records in a single sql
        "INSERT INTO `d1` values (now-8s, 10, 120, 0.33) (now - 6s, 10, 119, 0.34) (now - 4s, 11.2, 118, 0.322)",
    ]).await?;

    assert_eq!(inserted, 6);
    let mut result = taos.query("select * from `meters`").await?;

    for field in result.fields() {
        println!("got field: {}", field.name());
    }

    // Query option 1, use rows stream.
    let mut rows = result.rows();
    while let Some(row) = rows.try_next().await? {
        for (name, value) in row {
            println!("got value of {}: {}", name, value);
        }
    }

    // Query options 2, use deserialization with serde.
    #[derive(Debug, serde::Deserialize)]
    #[allow(dead_code)]
    struct Record {
        // deserialize timestamp to chrono::DateTime<Local>
        ts: DateTime<Local>,
        // float to f32
        current: Option<f32>,
        // int to i32
        voltage: Option<i32>,
        phase: Option<f32>,
        groupid: i32,
        // binary/varchar to String
        location: String,
    }

    let records: Vec<Record> = taos
        .query("select * from `meters`")
        .await?
        .deserialize()
        .try_collect()
        .await?;

    dbg!(records);
    Ok(())
}
```

### Subscription

```rust
use std::time::Duration;

use chrono::{DateTime, Local};
use taos::*;

// Query options 2, use deserialization with serde.
#[derive(Debug, serde::Deserialize)]
#[allow(dead_code)]
struct Record {
    // deserialize timestamp to chrono::DateTime<Local>
    ts: DateTime<Local>,
    // float to f32
    current: Option<f32>,
    // int to i32
    voltage: Option<i32>,
    phase: Option<f32>,
}

async fn prepare(taos: Taos) -> anyhow::Result<()> {
    let inserted = taos.exec_many([
        // create child table
        "CREATE TABLE `d0` USING `meters` TAGS(0, 'Los Angles')",
        // insert into child table
        "INSERT INTO `d0` values(now - 10s, 10, 116, 0.32)",
        // insert with NULL values
        "INSERT INTO `d0` values(now - 8s, NULL, NULL, NULL)",
        // insert and automatically create table with tags if not exists
        "INSERT INTO `d1` USING `meters` TAGS(1, 'San Francisco') values(now - 9s, 10.1, 119, 0.33)",
        // insert many records in a single sql
        "INSERT INTO `d1` values (now-8s, 10, 120, 0.33) (now - 6s, 10, 119, 0.34) (now - 4s, 11.2, 118, 0.322)",
    ]).await?;
    assert_eq!(inserted, 6);
    Ok(())
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // std::env::set_var("RUST_LOG", "debug");
    pretty_env_logger::init();
    let dsn = "taos://localhost:6030";
    let builder = TaosBuilder::from_dsn(dsn)?;

    let taos = builder.build()?;
    let db = "tmq";

    // prepare database
    taos.exec_many([
        format!("DROP TOPIC IF EXISTS tmq_meters"),
        format!("DROP DATABASE IF EXISTS `{db}`"),
        format!("CREATE DATABASE `{db}`"),
        format!("USE `{db}`"),
        // create super table
        format!("CREATE TABLE `meters` (`ts` TIMESTAMP, `current` FLOAT, `voltage` INT, `phase` FLOAT)\
                 TAGS (`groupid` INT, `location` BINARY(16))"),
        // create topic for subscription
        format!("CREATE TOPIC tmq_meters with META AS DATABASE {db}")
    ])
    .await?;

    let task = tokio::spawn(prepare(taos));

    tokio::time::sleep(Duration::from_secs(1)).await;

    // subscribe
    let tmq = TmqBuilder::from_dsn("taos://localhost:6030/?group.id=test")?;

    let mut consumer = tmq.build()?;
    consumer.subscribe(["tmq_meters"]).await?;

    {
        let mut stream = consumer.stream();

        while let Some((offset, message)) = stream.try_next().await? {
            // get information from offset

            // the topic
            let topic = offset.topic();
            // the vgroup id, like partition id in kafka.
            let vgroup_id = offset.vgroup_id();
            println!("* in vgroup id {vgroup_id} of topic {topic}\n");

            if let Some(data) = message.into_data() {
                while let Some(block) = data.fetch_raw_block().await? {
                    // one block for one table, get table name if needed
                    let name = block.table_name();
                    let records: Vec<Record> = block.deserialize().try_collect()?;
                    println!(
                        "** table: {}, got {} records: {:#?}\n",
                        name.unwrap(),
                        records.len(),
                        records
                    );
                }
            }
            consumer.commit(offset).await?;
        }
    }

    consumer.unsubscribe().await;

    task.await??;

    Ok(())
}

```

## Contribution

Welcome for all contributions.

## License

Keep same with [TDengine].

[TDengine]: https://www.taosdata.com/en/getting-started/
[r2d2]: https://crates.io/crates/r2d2


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [taos-v0.9.4] - 2023-08-09

**Full Changelog**: [taos-v0.9.3...taos-v0.9.4](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.9.3...taos-v0.9.4)

### Bug Fixes

- *ws*: Fix some panics in websocket impls ([142a9e9](142a9e9e8806cd3425568179d0afbe0c23ea2487))



### Features


- Use default rustls for taos-ws-sys ([52e9b18](52e9b186a39cd13e15462b46621f89fbb40c148c))


## [taos-v0.9.3] - 2023-08-03

**Full Changelog**: [taos-v0.9.2...taos-v0.9.3](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.9.2...taos-v0.9.3)

### Bug Fixes

- *mdsn*: Fix unix path with spaces ([d5aa25f](d5aa25f3dd3ff07b29d90fa9b4d281739eb549ce))



### Documentation


- Add example code ([0cb8868](0cb886841f68c4af8c862c4c26a85a7c3be3cc29))
- Add example code: basic ([01788f9](01788f96a5d21d2faacd18636f977f3302af3b19))
- Add example code: stmt bind ([1dcf95b](1dcf95bd1ca8c4b3b85f5168d7718afc12ddc0f3))
- Add example code: stmt muti bind ([ed81b2d](ed81b2d8228b04886b30846847c900a5511f316a))
- Add example code: stmt muti bind ([c9b64fd](c9b64fd7b5e8565ef25ec705fe0509f6d7652422))
- Add ws-table-muti-bind ([69d4054](69d4054c42ac37de52f05561e3440434e70af833))
- Add example of get err string ([56400af](56400af2c3ba3ddfce6c128dbd41b80575e30a2f))


## [taos-v0.9.2] - 2023-07-27

**Full Changelog**: [taos-v0.9.1...taos-v0.9.2](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.9.1...taos-v0.9.2)

### Bug Fixes

- *sys*: Fix nchar display error ([2c291f1](2c291f1ceed52f546e08dad5d3efabc8caaf8cfb))
- *ws-sys*: Fix nchar to str ([00c465c](00c465c939717060a288cd8c59f563429f3fc35f))

- Reclaim vec item ([6ad6f81](6ad6f811db820be19e4286eb34396d4a259311d1))
- Fix csv dsn parse error when use unix glob ([f49a0e5](f49a0e595309a5bd5a4fc12c632f48a6bbeeacd8))


### Documentation


- Update git cliff config and CHANGELOG.md ([c8848c9](c8848c9b87febe0b4f3cab044f310c968d815052))


### Enhancements


- Add ws_stmt_reclaim_fields ([cb52f6e](cb52f6e9a38e47051387f586b87a5629a6554f06))
- Add error handler ([9e18dfa](9e18dfaafec4e435e99db125d737b9cc0fefe318))


### Features


- Add api get_col_fields and get_tag_fields ([83c0242](83c0242a080e9ef3b5b9f0d20e8581c7f3a1d9ae))
- Add c api stmt get fields ([552749d](552749d87053d69ea53d14e0ce657d05ef6b355d))
- Update C struct ([fdebcd7](fdebcd7767779da546596e207b4829a7e20596cf))


### Testing


- Add case for taos-ws-sys ([0624b6f](0624b6fdcc889b4c442cb164276e1d7ff8322a7f))
- Update case ([f1db56b](f1db56bf747ddb2028c719c9876685b201c9d965))
- Update case ([bb6f1d1](bb6f1d1f145763e82d29b1a447c00c3582e9e38d))


## [taos-v0.9.1] - 2023-07-18

**Full Changelog**: [taos-v0.9.0...taos-v0.9.1](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.9.0...taos-v0.9.1)

### Bug Fixes


- Table does not exist in enterprise checking #[TD-25253](https://jira.taosdata.com:18080/browse/TD-25253) ([316945f](316945fb7b261e3cec24f1b148fb42ebe0a4e8ad))


## [taos-v0.9.0] - 2023-07-12

**Full Changelog**: [taos-v0.8.15...taos-v0.9.0](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.8.15...taos-v0.9.0)

### Bug Fixes

- *optin*: Safe way to fetch raw block in both v2/v3 ([931a3f5](931a3f5cda145919f54b394d756388f66931f49a))

- Annotate of ws_get_server_info ([1d05e7b](1d05e7b1ad9e103791d055aaa1d8d790d4128861))
- Annotate of ws_get_server_info ([295cf45](295cf4535761e0e0d2d21b13508910bd7682b08e))
- Drop data in WsMaybeError ([db84de5](db84de5271bce8ec7d96ec3562748af958341cd9))
- Remove read ([4a4cdac](4a4cdac2ea020ce3ab834d4da8778ef95b0103f3))
- Drop data in WsMaybeError ([ced57b9](ced57b963f24d317567cee2425831af053842cf8))
- Remove read ([a40f0a8](a40f0a81e3ff9c146b7f875ce548a76848197bb2))
- Fix unexpected error string ([3e79cf1](3e79cf1db41ad034ad97b28c5fecaa6872fb253d))


### Enhancements


- Improve query errors ([2f4d84b](2f4d84b5f5d81f3a5055c890c9cb0c481b5dcbaf))


### Features


- Raise backtrace with taosc error codes ([2637b39](2637b39b4ab1269b543c6ee7203eca445b3c9bf4))


### Testing


- Add benchmark ([ee9a867](ee9a8675e2fef06501fea8e025a4bf798d5992cb))
- Rename bench file name ([c94e70c](c94e70cfba4cba35ec518da5d8881cdcd89bf218))
- Update case ([a610057](a610057ee627029443a8827a713c39aeffa2438b))
- Add bench reports ([9ce79e2](9ce79e2c6f1416e070ac75fcf2c3d928b87fd874))


## [taos-v0.8.15] - 2023-06-28

**Full Changelog**: [taos-v0.8.14...taos-v0.8.15](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.8.14...taos-v0.8.15)

### Bug Fixes


- Cloud as enterprise edition ([f93c339](f93c339f53498ee66573a010429c1dfd0470f8e9))


## [taos-v0.8.14] - 2023-06-27

**Full Changelog**: [taos-v0.8.13...taos-v0.8.14](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.8.13...taos-v0.8.14)

### Bug Fixes


- Treat cloud as enterprise ([68bdda9](68bdda9ec22d3cbd1ae1057d4dddf2e64577fd8a))


## [taos-v0.8.13] - 2023-06-27

**Full Changelog**: [taos-v0.8.12...taos-v0.8.13](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.8.12...taos-v0.8.13)

### Bug Fixes


- Cloud connection 400 bad request error ([6281978](6281978ca7933ec4965dd74ecdb958fe1185d00b))


## [taos-v0.8.12] - 2023-06-21

**Full Changelog**: [taos-v0.8.11...taos-v0.8.12](https://github.com/taosdata/taos-connector-rust/compare/taos-v0.8.11...taos-v0.8.12)

### Bug Fixes


- Update free result ([36a7b41](36a7b41fcbe5e289dba6b54fb7ade603e373b857))
- Add thread scope ([5b37e63](5b37e63f51ee65470f87294ade500bd97178c438))
- Retry connect when 0x000B error ([5a885ab](5a885abb7c2160ec124d745a33ebfd9c859e8c0a))


### Enhancements


- Add sync implement for put schemaless ([5a57e5d](5a57e5d5ec05c5879d8da49ec2bf21d1ecdeff46))


## [taos-v0.8.11] - 2023-06-13


src/query.rs
use super::*;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error(transparent)]
    Dsn(#[from] DsnError),
    #[error(transparent)]
    Raw(#[from] RawError),
    #[error(transparent)]
    Native(#[from] taos_sys::Error),
    #[error(transparent)]
    Ws(#[from] taos_ws::Error),
    #[error(transparent)]
    WsQueryError(#[from] taos_ws::query::asyn::Error),
    #[error(transparent)]
    WsTmqError(#[from] taos_ws::consumer::Error),
    #[error(transparent)]
    Any(#[from] anyhow::Error),
}

enum TaosBuilderInner {
    Native(taos_sys::TaosBuilder),
    Ws(taos_ws::TaosBuilder),
}
pub(super) enum TaosInner {
    Native(taos_sys::Taos),
    Ws(taos_ws::Taos),
}

enum ResultSetInner {
    Native(taos_sys::ResultSet),
    Ws(taos_ws::ResultSet),
}

pub struct TaosBuilder(TaosBuilderInner);
pub struct Taos(pub(super) TaosInner);
pub struct ResultSet(ResultSetInner);

impl TBuilder for TaosBuilder {
    type Target = Taos;

    type Error = Error;

    fn available_params() -> &'static [&'static str] {
        &[]
    }

    fn from_dsn<D: IntoDsn>(dsn: D) -> Result<Self, Self::Error> {
        let dsn = dsn.into_dsn()?;
        // dbg!(&dsn);
        match (dsn.driver.as_str(), dsn.protocol.as_deref()) {
            ("ws" | "wss" | "http" | "https" | "taosws" | "taoswss", _) => Ok(Self(
                TaosBuilderInner::Ws(taos_ws::TaosBuilder::from_dsn(dsn)?),
            )),
            ("taos" | "tmq", None) => Ok(Self(TaosBuilderInner::Native(
                taos_sys::TaosBuilder::from_dsn(dsn)?,
            ))),
            ("taos" | "tmq", Some("ws" | "wss" | "http" | "https")) => Ok(Self(
                TaosBuilderInner::Ws(taos_ws::TaosBuilder::from_dsn(dsn)?),
            )),
            (driver, _) => Err(DsnError::InvalidDriver(driver.to_string()).into()),
        }
    }

    fn client_version() -> &'static str {
        ""
    }

    fn ping(&self, conn: &mut Self::Target) -> Result<(), Self::Error> {
        match &self.0 {
            TaosBuilderInner::Native(b) => match &mut conn.0 {
                TaosInner::Native(taos) => Ok(b.ping(taos)?),
                _ => unreachable!(),
            },
            TaosBuilderInner::Ws(b) => match &mut conn.0 {
                TaosInner::Ws(taos) => Ok(b.ping(taos)?),
                _ => unreachable!(),
            },
        }
    }

    fn ready(&self) -> bool {
        match &self.0 {
            TaosBuilderInner::Native(b) => b.ready(),
            TaosBuilderInner::Ws(b) => b.ready(),
        }
    }

    fn build(&self) -> Result<Self::Target, Self::Error> {
        match &self.0 {
            TaosBuilderInner::Native(b) => Ok(Taos(TaosInner::Native(b.build()?))),
            TaosBuilderInner::Ws(b) => Ok(Taos(TaosInner::Ws(b.build()?))),
        }
    }
}

impl AsyncFetchable for ResultSet {
    type Error = Error;

    fn affected_rows(&self) -> i32 {
        match &self.0 {
            ResultSetInner::Native(rs) => {
                <taos_sys::ResultSet as AsyncFetchable>::affected_rows(rs)
            }
            ResultSetInner::Ws(rs) => <taos_ws::ResultSet as AsyncFetchable>::affected_rows(rs),
        }
    }

    fn precision(&self) -> Precision {
        match &self.0 {
            ResultSetInner::Native(rs) => <taos_sys::ResultSet as AsyncFetchable>::precision(rs),
            ResultSetInner::Ws(rs) => <taos_ws::ResultSet as AsyncFetchable>::precision(rs),
        }
    }

    fn fields(&self) -> &[Field] {
        match &self.0 {
            ResultSetInner::Native(rs) => <taos_sys::ResultSet as AsyncFetchable>::fields(rs),
            ResultSetInner::Ws(rs) => <taos_ws::ResultSet as AsyncFetchable>::fields(rs),
        }
    }

    fn summary(&self) -> (usize, usize) {
        match &self.0 {
            ResultSetInner::Native(rs) => <taos_sys::ResultSet as AsyncFetchable>::summary(rs),
            ResultSetInner::Ws(rs) => <taos_ws::ResultSet as AsyncFetchable>::summary(rs),
        }
    }

    fn update_summary(&mut self, nrows: usize) {
        match &mut self.0 {
            ResultSetInner::Native(rs) => {
                <taos_sys::ResultSet as AsyncFetchable>::update_summary(rs, nrows)
            }
            ResultSetInner::Ws(rs) => {
                <taos_ws::ResultSet as AsyncFetchable>::update_summary(rs, nrows)
            }
        }
    }

    fn fetch_raw_block(
        &mut self,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Result<Option<RawBlock>, Self::Error>> {
        match &mut self.0 {
            ResultSetInner::Native(rs) => {
                <taos_sys::ResultSet as AsyncFetchable>::fetch_raw_block(rs, cx).map_err(Into::into)
            }
            ResultSetInner::Ws(rs) => {
                <taos_ws::ResultSet as AsyncFetchable>::fetch_raw_block(rs, cx).map_err(Into::into)
            }
        }
    }
}

impl taos_query::Fetchable for ResultSet {
    type Error = Error;

    fn affected_rows(&self) -> i32 {
        match &self.0 {
            ResultSetInner::Native(rs) => {
                <taos_sys::ResultSet as AsyncFetchable>::affected_rows(rs)
            }
            ResultSetInner::Ws(rs) => <taos_ws::ResultSet as AsyncFetchable>::affected_rows(rs),
        }
    }

    fn precision(&self) -> Precision {
        match &self.0 {
            ResultSetInner::Native(rs) => <taos_sys::ResultSet as AsyncFetchable>::precision(rs),
            ResultSetInner::Ws(rs) => <taos_ws::ResultSet as AsyncFetchable>::precision(rs),
        }
    }

    fn fields(&self) -> &[Field] {
        match &self.0 {
            ResultSetInner::Native(rs) => <taos_sys::ResultSet as AsyncFetchable>::fields(rs),
            ResultSetInner::Ws(rs) => <taos_ws::ResultSet as AsyncFetchable>::fields(rs),
        }
    }

    fn summary(&self) -> (usize, usize) {
        match &self.0 {
            ResultSetInner::Native(rs) => <taos_sys::ResultSet as AsyncFetchable>::summary(rs),
            ResultSetInner::Ws(rs) => <taos_ws::ResultSet as AsyncFetchable>::summary(rs),
        }
    }

    fn update_summary(&mut self, nrows: usize) {
        match &mut self.0 {
            ResultSetInner::Native(rs) => {
                <taos_sys::ResultSet as AsyncFetchable>::update_summary(rs, nrows)
            }
            ResultSetInner::Ws(rs) => {
                <taos_ws::ResultSet as AsyncFetchable>::update_summary(rs, nrows)
            }
        }
    }

    fn fetch_raw_block(&mut self) -> Result<Option<RawBlock>, Self::Error> {
        match &mut self.0 {
            ResultSetInner::Native(rs) => {
                <taos_sys::ResultSet as taos_query::Fetchable>::fetch_raw_block(rs)
                    .map_err(Into::into)
            }
            ResultSetInner::Ws(rs) => {
                <taos_ws::ResultSet as taos_query::Fetchable>::fetch_raw_block(rs)
                    .map_err(Into::into)
            }
        }
    }
}

#[async_trait::async_trait]
impl AsyncQueryable for Taos {
    type Error = Error;

    type AsyncResultSet = ResultSet;

    async fn query<T: AsRef<str> + Send + Sync>(
        &self,
        sql: T,
    ) -> Result<Self::AsyncResultSet, Self::Error> {
        match &self.0 {
            TaosInner::Native(taos) => taos
                .query(sql)
                .await
                .map(ResultSetInner::Native)
                .map(ResultSet)
                .map_err(Into::into),
            TaosInner::Ws(taos) => taos
                .query(sql)
                .await
                .map(ResultSetInner::Ws)
                .map(ResultSet)
                .map_err(Into::into),
        }
    }

    async fn write_raw_meta(&self, meta: RawMeta) -> Result<(), Self::Error> {
        match &self.0 {
            TaosInner::Native(taos) => taos.write_raw_meta(meta).await.map_err(Into::into),
            TaosInner::Ws(taos) => taos.write_raw_meta(meta).await.map_err(Into::into),
        }
    }

    async fn write_raw_block(&self, block: &RawBlock) -> Result<(), Self::Error> {
        match &self.0 {
            TaosInner::Native(taos) => taos.write_raw_block(block).await.map_err(Into::into),
            TaosInner::Ws(taos) => taos.write_raw_block(block).await.map_err(Into::into),
        }
    }
}

impl taos_query::Queryable for Taos {
    type Error = Error;

    type ResultSet = ResultSet;

    fn query<T: AsRef<str>>(&self, sql: T) -> Result<Self::ResultSet, Self::Error> {
        match &self.0 {
            TaosInner::Native(taos) => <taos_sys::Taos as taos_query::Queryable>::query(taos, sql)
                .map(ResultSetInner::Native)
                .map(ResultSet)
                .map_err(Into::into),
            TaosInner::Ws(taos) => <taos_ws::Taos as taos_query::Queryable>::query(taos, sql)
                .map(ResultSetInner::Ws)
                .map(ResultSet)
                .map_err(Into::into),
        }
    }

    fn write_meta(&self, meta: RawMeta) -> Result<(), Self::Error> {
        match &self.0 {
            TaosInner::Native(taos) => {
                <taos_sys::Taos as taos_query::Queryable>::write_meta(taos, meta)
                    .map_err(Into::into)
            }
            TaosInner::Ws(taos) => {
                <taos_ws::Taos as taos_query::Queryable>::write_meta(taos, meta).map_err(Into::into)
            }
        }
    }
}
#[cfg(test)]
mod tests {

    use std::str::FromStr;

    use taos_query::{common::Timestamp, TBuilder};

    use super::TaosBuilder;

    #[test]
    fn builder() {
        let builder = TaosBuilder::from_dsn("taos://").unwrap();
        assert!(builder.ready());

        let mut conn = builder.build().unwrap();
        assert!(builder.ping(&mut conn).is_ok());
    }

    #[test]
    fn sync_json_test_native() -> anyhow::Result<()> {
        let dsn = std::env::var("TEST_DSN").unwrap_or("taos://".to_string());
        sync_json_test(&dsn, "taos")
    }
    #[cfg(feature = "ws")]
    #[test]
    fn sync_json_test_ws() -> anyhow::Result<()> {
        sync_json_test("ws://", "ws")
    }
    #[cfg(feature = "ws")]
    #[test]
    fn sync_json_test_taosws() -> anyhow::Result<()> {
        sync_json_test("taosws://", "taosws")
    }

    #[test]
    fn null_test() -> anyhow::Result<()> {
        use taos_query::prelude::sync::*;
        let dsn = std::env::var("TEST_DSN").unwrap_or("taos://localhost:6030".to_string());
        let dsn = Dsn::from_str(&dsn)?;
        let taos = TaosBuilder::from_dsn(&dsn)?.build()?;
        taos.exec_many(["drop database if exists db", "create database db", "use db"])?;

        taos.exec(
            "create table st(ts timestamp, c1 TINYINT UNSIGNED) tags(utntag TINYINT UNSIGNED)",
        )?;
        taos.exec("create table t1 using st tags(0)")?;
        taos.exec("insert into t1 values(1640000000000, 0)")?;
        taos.exec("create table t2 using st tags(254)")?;
        taos.exec("insert into t2 values(1640000000000, 254)")?;
        taos.exec("create table t3 using st tags(NULL)")?;
        taos.exec("insert into t3 values(1640000000000, NULL)")?;

        let mut rs = taos.query("select * from st where utntag is null")?;
        for row in rs.rows() {
            let row = row?;
            let values = row.into_values();
            assert_eq!(values[1], Value::Null(Ty::UTinyInt));
            assert_eq!(values[2], Value::Null(Ty::UTinyInt));
        }
        Ok(())
    }

    fn sync_json_test(dsn: &str, db: &str) -> anyhow::Result<()> {
        use taos_query::prelude::sync::*;

        std::env::set_var("RUST_LOG", "debug");
        // pretty_env_logger::init();
        use taos_query::{Fetchable, Queryable};
        let client = TaosBuilder::from_dsn(dsn)?.build()?;
        assert_eq!(client.exec(format!("drop database if exists {db}"))?, 0);
        assert_eq!(client.exec(format!("create database {db} keep 36500"))?, 0);
        assert_eq!(
            client.exec(
                format!("create table {db}.stb1(ts timestamp,\
                    b1 bool, c8i1 tinyint, c16i1 smallint, c32i1 int, c64i1 bigint,\
                    c8u1 tinyint unsigned, c16u1 smallint unsigned, c32u1 int unsigned, c64u1 bigint unsigned,\
                    cb1 binary(100), cn1 nchar(10),

                    b2 bool, c8i2 tinyint, c16i2 smallint, c32i2 int, c64i2 bigint,\
                    c8u2 tinyint unsigned, c16u2 smallint unsigned, c32u2 int unsigned, c64u2 bigint unsigned,\
                    cb2 binary(10), cn2 nchar(16)) tags (jt json)")
            )?,
            0
        );
        assert_eq!(
            client.exec(format!(
                r#"insert into {db}.tb1 using {db}.stb1 tags('{{"key":"数据"}}')
                   values(0,    true, -1,  -2,  -3,  -4,   1,   2,   3,   4,   'abc', '涛思',
                                false,-5,  -6,  -7,  -8,   5,   6,   7,   8,   'def', '数据')
                         (65535,NULL, NULL,NULL,NULL,NULL, NULL,NULL,NULL,NULL, NULL,  NULL,
                                NULL, NULL,NULL,NULL,NULL, NULL,NULL,NULL,NULL, NULL,  NULL)"#
            ))?,
            2
        );
        assert_eq!(
            client.exec(format!(
                r#"insert into {db}.tb2 using {db}.stb1 tags(NULL)
                   values(1,    true, -1,  -2,  -3,  -4,   1,   2,   3,   4,   'abc', '涛思',
                                false,-5,  -6,  -7,  -8,   5,   6,   7,   8,   'def', '数据')
                         (65536,NULL, NULL,NULL,NULL,NULL, NULL,NULL,NULL,NULL, NULL,  NULL,
                                NULL, NULL,NULL,NULL,NULL, NULL,NULL,NULL,NULL, NULL,  NULL)"#
            ))?,
            2
        );

        // let mut rs = client.s_query("select * from wsabc.tb1").unwrap().unwrap();
        let mut rs = client.query(format!("select * from {db}.tb1 order by ts limit 1"))?;

        #[derive(Debug, serde::Deserialize, PartialEq, Eq)]
        #[allow(dead_code)]
        struct A {
            ts: String,
            b1: bool,
            c8i1: i8,
            c16i1: i16,
            c32i1: i32,
            c64i1: i64,
            c8u1: u8,
            c16u1: u16,
            c32u1: u32,
            c64u1: u64,

            c8i2: i8,
            c16i2: i16,
            c32i2: i32,
            c64i2: i64,
            c8u2: u8,
            c16u2: u16,
            c32u2: u32,
            c64u2: u64,

            cb1: String,
            cb2: String,
            cn1: String,
            cn2: String,
        }

        use itertools::Itertools;
        let values: Vec<A> = rs.deserialize::<A>().try_collect()?;

        dbg!(&values);

        assert_eq!(
            values[0],
            A {
                ts: format!(
                    "{}",
                    Value::Timestamp(Timestamp::new(0, Precision::Millisecond))
                ),
                b1: true,
                c8i1: -1,
                c16i1: -2,
                c32i1: -3,
                c64i1: -4,
                c8u1: 1,
                c16u1: 2,
                c32u1: 3,
                c64u1: 4,
                c8i2: -5,
                c16i2: -6,
                c32i2: -7,
                c64i2: -8,
                c8u2: 5,
                c16u2: 6,
                c32u2: 7,
                c64u2: 8,
                cb1: "abc".to_string(),
                cb2: "def".to_string(),
                cn1: "涛思".to_string(),
                cn2: "数据".to_string(),
            }
        );

        assert_eq!(client.exec(format!("drop database {db}"))?, 0);
        Ok(())
    }
}
